src/newton_solver.cpp

```diff
@@ -410,38 +368,150 @@ int NewtonSolverIterative::prepareLinearSystem(int ntry, int nnewt) {
+    N_VConst(0.0, ns_delta);
+    N_VConst(0.0, ns_tmp);
+    rho = 1.0;
+    omega = 1.0;
+    alpha = 1.0;
+    
+    // can be set to 0 at the moment
+    model->fJv(tdata->t, tdata->x, tdata->dx, tdata->xdot, ns_delta, ns_Jv, 0.0,  tdata, ns_tmp, NULL);
+    
+    // ns_r = xdot - ns_Jv;
+    N_VLinearSum(-1.0, ns_Jv, 1.0, tdata->xdot, ns_r);
+    N_VDiv(ns_r, ns_Jdiag, ns_r);
+    res = sqrt(N_VDotProd(ns_r, ns_r));
+    N_VScale(1.0, ns_r, ns_rt);
+    
+    for (int i_linstep = 0; i_linstep < udata->newton_maxlinsteps;
+         i_linstep++) {
+        // Compute factors
+        rho1 = rho;
+        rho = N_VDotProd(ns_rt, ns_r);
+        beta = rho * alpha / (rho1 * omega);
+        
+        // ns_p = ns_r + beta * (ns_p - omega * ns_v);
+        N_VLinearSum(1.0, ns_p, -omega, ns_v, ns_p);
+        N_VLinearSum(1.0, ns_r, beta, ns_p, ns_p);
+        
+        // ns_v = J * ns_p
+        model->fJv(tdata->t, tdata->x, tdata->dx, tdata->xdot, ns_p, ns_v, 0.0,  tdata, ns_tmp, NULL);
+        N_VDiv(ns_v, ns_Jdiag, ns_v);
+        
+        // Compute factor
+        alpha = rho / N_VDotProd(ns_rt, ns_v);
+        
+        // ns_h = ns_delta + alpha * ns_p;
+        N_VLinearSum(1.0, ns_delta, alpha, ns_p, ns_h);
+        // ns_s = ns_r - alpha * ns_v;
+        N_VLinearSum(1.0, ns_r, -alpha, ns_v, ns_s);
+        
+        // ns_t = J * ns_s
+        model->fJv(tdata->t, tdata->x, tdata->dx, tdata->xdot, ns_s, ns_t, 0.0,  tdata, ns_tmp, NULL);
+        N_VDiv(ns_t, ns_Jdiag, ns_t);
+        
+        // Compute factor
+        omega = N_VDotProd(ns_t, ns_s) / N_VDotProd(ns_t, ns_t);
+        
+        // ns_delta = ns_h + omega * ns_s;
+        N_VLinearSum(1.0, ns_h, omega, ns_s, ns_delta);
+        // ns_r = ns_s - omega * ns_t;
+        N_VLinearSum(1.0, ns_s, -omega, ns_t, ns_r);
+        
+        // Compute the (unscaled) residual
+        N_VProd(ns_r, ns_Jdiag, ns_r);
+        res = sqrt(N_VDotProd(ns_r, ns_r));
+        
+        // Test convergence
+        if (res < udata->atol) {
+            // Write number of steps needed
+            rdata->newton_numlinsteps[(ntry - 1) * udata->newton_maxsteps +
+                                      nnewt] = i_linstep + 1;
+            
+            // Return success
+            return;
+        }
+        
+        // Scale back
+        N_VDiv(ns_r, ns_Jdiag, ns_r);
+    }
+    throw NewtonFailure("SPBCG solver failed to converge");
 }
+    
 
 /* ----------------------------------------------------------------------------------
  */
 
-NewtonSolverIterative::~NewtonSolverIterative(){};
+NewtonSolverIterative::~NewtonSolverIterative(){
+    N_VDestroy_Serial(ns_p);
+    N_VDestroy_Serial(ns_h);
+    N_VDestroy_Serial(ns_t);
+    N_VDestroy_Serial(ns_s);
+    N_VDestroy_Serial(ns_r);
+    N_VDestroy_Serial(ns_rt);
+    N_VDestroy_Serial(ns_v);
+    N_VDestroy_Serial(ns_Jv);
+    N_VDestroy_Serial(ns_tmp);
+    N_VDestroy_Serial(ns_Jdiag);
+    N_VScale(-1.0, tdata->xdot, tdata->xdot);
```

Comment:
This is asking for trouble... Depending on when the object is destroyed, the scaling might happen at the wrong time.